src/main/cms/cms.c

```diff
@@ -0,0 +1,848 @@
+#include <stdint.h>
+#include <string.h>
+#include <ctype.h>
+
+#include "platform.h"
+
+#ifdef CMS
+
+#include "build/version.h"
+#include "build/debug.h"
+
+#include "cms/cms.h"
+#include "cms/cms_menu_builtin.h"
+#include "cms/cms_types.h"
+
+#include "common/typeconversion.h"
+
+#include "drivers/system.h"
+
+// For 'ARM' related
+#include "fc/config.h"
+#include "fc/rc_controls.h"
+#include "fc/runtime_config.h"
+
+// For rcData, stopAllMotors, stopPwmAllMotors
+#include "config/config_profile.h"
+#include "config/config_master.h"
+#include "config/feature.h"
+
+// For VISIBLE* (Actually, included by config_master.h)
+#include "io/osd.h"
+
+// DisplayPort management
+
+#ifndef CMS_MAX_DEVICE
+#define CMS_MAX_DEVICE 4
+#endif
+
+static displayPort_t *pCurrentDisplay;
+
+static displayPort_t *cmsDisplayPorts[CMS_MAX_DEVICE];
+static int cmsDeviceCount;
+static int cmsCurrentDevice = -1;
+
+bool cmsDisplayPortRegister(displayPort_t *pDisplay)
+{
+    if (cmsDeviceCount == CMS_MAX_DEVICE)
+        return false;
+
+    cmsDisplayPorts[cmsDeviceCount++] = pDisplay;
+
+    return true;
+}
+
+static displayPort_t *cmsDisplayPortSelectCurrent(void)
+{
+    if (cmsDeviceCount == 0)
+        return NULL;
+
+    if (cmsCurrentDevice < 0)
+        cmsCurrentDevice = 0;
+
+    return cmsDisplayPorts[cmsCurrentDevice];
+}
+
+static displayPort_t *cmsDisplayPortSelectNext(void)
+{
+    if (cmsDeviceCount == 0)
+        return NULL;
+
+    cmsCurrentDevice = (cmsCurrentDevice + 1) % cmsDeviceCount; // -1 Okay
+
+    return cmsDisplayPorts[cmsCurrentDevice];
+}
+
+#define CMS_UPDATE_INTERVAL  50   // Interval of key scans (msec)
+#define CMS_POLL_INTERVAL   100   // Interval of polling dynamic values (msec)
+
+// XXX LEFT_MENU_COLUMN and RIGHT_MENU_COLUMN must be adjusted
+// dynamically depending on size of the active output device,
+// or statically to accomodate sizes of all supported devices.
+//
+// Device characteristics
+// OLED
+//   21 cols x 8 rows
+//     128x64 with 5x7 (6x8) : 21 cols x 8 rows
+// MAX7456 (PAL)
+//   30 cols x 16 rows
+// MAX7456 (NTSC)
+//   30 cols x 13 rows
+// HoTT Telemetry Screen
+//   21 cols x 8 rows
+//
+
+#define LEFT_MENU_COLUMN  1
+#define RIGHT_MENU_COLUMN(p) ((p)->cols - 8)
+#define MAX_MENU_ITEMS(p)    ((p)->rows - 2)
+
+static bool cmsInMenu = false;
+
+static CMS_Menu *currentMenu;    // Points to top entry of the current page
+
+// XXX Does menu backing support backing into second page???
+
+static CMS_Menu *menuStack[10];  // Stack to save menu transition
+static uint8_t menuStackHistory[10];// cursorRow in a stacked menu
+static uint8_t menuStackIdx = 0;
+
+static OSD_Entry *pageTop;       // Points to top entry of the current page
+static OSD_Entry *pageTopAlt;    // Only 2 pages are allowed (for now)
+static uint8_t maxRow;           // Max row in the current page
+
+static int8_t cursorRow;
+
+#ifdef CMS_MENU_DEBUG // For external menu content creators
+
+static char menuErrLabel[21 + 1] = "RAMDOM DATA";
```

Comment:
Yikes! And yes, it could be something more intuitive, but users will never see it anyway.